ols/src/llms/llm_loader.py

```diff
@@ -54,44 +46,55 @@ class LLMLoader:
         params = {'temperature': 0.02, 'top_p': 0.95}
 
         bare_llm = LLMLoader(provider="openai", model="gpt-3.5-turbo", params=params)
-        llm_chain = LLMChain(llm=bare_llm, prompt=prompt)
+        llm_chain = LLMChain(llm=bare_llm.llm, prompt=prompt)
         ```
     """
 
     def __init__(
         self,
-        provider: Optional[str] = None,
-        model: Optional[str] = None,
+        provider: str,
+        model: str,
         params: Optional[dict] = None,
+        llm_providers: Optional[LLMProviders] = None,
     ) -> None:
         """Initialize LLM loader.
 
         Args:
             provider: The LLM provider name.
             model: The LLM model name.
             params: The parameters to override and pass to the LLM backend.
+            llm_providers: The LLM providers configuration.
 
         Raises:
-            MissingProviderError: When provider is not specified.
-            MissingModelError: When model is not specified.
-            UnsupportedProviderError: When provider is not supported or is unknown.
+            UnknownProviderError: When provider is not is unknown.
+            UnsupportedProviderError: When provider is unsupported.
             ModelConfigMissingError: When no configuration exists for the requested model name.
         """
-        if provider is None:
-            msg = "Missing provider"
-            logger.error(msg)
-            raise MissingProviderError(msg)
         self.provider = provider
-        if model is None:
-            msg = "Missing model"
-            logger.error(msg)
-            raise MissingModelError(msg)
         self.model = model
-        self.provider_config = self._get_provider_config()
+        self.params = params
+        self.llm_providers = llm_providers or config.llm_config
```

Comment:
when/why would we expect a caller to need to override the global config?

Reply:
It decouples components.
I agree that sometimes the only obvious benefit is having a clean unit test without patches, but at the same time, I think [dependency injection](https://python-dependency-injector.ets-labs.org/introduction/di_in_python.html#what-is-dependency-injection) is a good pattern to follow.

Do you have concerns with this? Could it confuse someone?

Reply:
by allowing the config to be passed in, you force llm_loader to validate the config that was passed in(or trust the caller to have done so), instead of being able to rely on the config loading logic to have verified the config.

so that is the trade-off to consider, the centralization of validation logic.  

if the only expected use for this parameter is ease of writing tests, then i don't think it belongs here.  If we want to make it easier to manipulate the OLS config for testing purposes, we can do that.

Reply:
> you force llm_loader to validate the config that was passed

Sure, there are some exceptions, but generally, I think it is the caller's responsibility. Otherwise, you would need to validate everything everywhere.

TBH this is not true DI as we are not injecting dependencies in the true meaning of the word (that is more complicated and requires other libs that do it for you - a bit similar how the fixtures works in tests).
So usually DI means making things just a bit more configurable. Often times, the only obvious benefit is better testability.
Here, we are decoupling `LLMLoader` and `LLMProviders`. It still depends on it, but it is not internal parameter.

But if you think it doesn't make sense, I can drop it.

Reply:
> Sure, there are some exceptions, but generally, I think it is the caller's responsibility. Otherwise, you would need to validate everything everywhere.

making the caller do it turns into "validate everything everywhere" because you will have many potential callers(places from which you call the function), only one callee.  Callers tend to just make the call and then expect to catch an exception if the call was invalid.  A function should pretty much never just trust the arguments it is passed w/o validating them.

> So usually DI means making things just a bit more configurable. Often times, the only obvious benefit is better testability.

and the downside is that functions that take more arguments need more logic to handle reconciling those arguments, especially when they are optional.  And then more tests to ensure the handling is done correctly.

You can leave this one as is, but there is definitely a fine line between structuring code to make it easy to test, and making the code more complicated than it needs to be and risking introducing errors as a result of that (or confusing callers of the code into doing things we don't really want them doing, because we only made it possible to make it easier to write tests, not because we wanted people actually doing it)